File: Pandacap.LowLevel/Data/Relationships/Follow.fs
namespace Pandacap.Data

open System
open Pandacap.ActivityPub
open Pandacap.PlatformBadges

/// An ActivityPub actor who this Pandacap actor is following.
type Follow() =
    inherit RemoteActorRelationship()

    member val FollowGuid = Guid.Empty with get, set
    member val Accepted = false with get, set
    member val IgnoreImages = false with get, set
    member val IncludeImageShares = false with get, set
    member val IncludeTextShares = false with get, set

    interface IActivityPubFollow with
        member this.ActorId = this.ActorId

    interface IFollow with
        member this.Filtered =
            not this.IncludeImageShares
            || not this.IncludeTextShares
        member _.Platform = ActivityPub
        member this.LinkUrl = this.Url |> orString this.ActorId
        member this.IconUrl = this.IconUrl
        member this.Username = this.PreferredUsername |> orString this.ActorId
        member this.Url = this.ActorId

File: Pandacap.Database/Follow.cs
using Pandacap.UI.Badges;
using Pandacap.UI.Elements;

namespace Pandacap.Database
{
    /// <summary>
    /// An ActivityPub actor who this Pandacap actor is following.
    /// </summary>
    public class Follow : RemoteActorRelationship, IFollow
    {
        /// <summary>
        /// A random ID used in the ActivityPub ID of the Follow activity that was sent.
        /// Used to send an Undo when unfollowing the actor.
        /// </summary>
        public Guid FollowGuid { get; set; }

        /// <summary>
        /// Whether the remote actor has accepted the follow request.
        /// </summary>
        public bool Accepted { get; set; }

        /// <summary>
        /// Whether to ignore images from this actor's posts when adding them to the inbox (sending them to the Text Posts section).
        /// </summary>
        public bool IgnoreImages { get; set; }

        /// <summary>
        /// Whether to include this actor's reposts/boosts if they contain images.
        /// </summary>
        public bool IncludeImageShares { get; set; }

        /// <summary>
        /// Whether to include this actor's reposts/boosts if they do not contain images.
        /// </summary>
        public bool IncludeTextShares { get; set; }

        Badge IFollow.Badge => Badges.ActivityPub;

        string? IFollow.LinkUrl => Url ?? ActorId;

        string IFollow.Username => PreferredUsername ?? ActorId;

        bool IFollow.Filtered => !IncludeImageShares || !IncludeTextShares;
    }
}
